servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancerFactory.java

```diff
@@ -95,13 +120,99 @@ public Builder() {
+         *
+         * @param backgroundExecutor {@link Executor} on which to schedule health checking.
+         * @return {@code this}.
+         */
+        public RoundRobinLoadBalancerFactory.Builder<ResolvedAddress, C> backgroundExecutor(
+                Executor backgroundExecutor) {
+            this.backgroundExecutor = requireNonNull(backgroundExecutor);
+            return this;
+        }
+
+        /**
+         * Configure an interval for health checking a host that failed to open connections.
+         * @param interval interval at which a background health check will be scheduled.
+         * @return {@code this}.
+         */
+        public RoundRobinLoadBalancerFactory.Builder<ResolvedAddress, C> healthCheckInterval(Duration interval) {
+            this.healthCheckInterval = requireNonNull(interval);
+            if (interval.isNegative()) {
+                throw new IllegalArgumentException("Health check interval can't be negative");
+            }
+            return this;
+        }
+
+        /**
+         * Configure a threshold for consecutive connection failures to a host. When the {@link LoadBalancer} fails
+         * to open a connection in more consecutive attempts than the specified value, the host will be marked as
+         * unhealthy and a connection establishment will take place in the background. Until finished, the host will
+         * not take part in load balancing selection.
+         * Use a negative value of the argument to disable health checking.
+         * @param threshold number of consecutive connection failures to consider a host unhealthy and eligible for
+         * background health checking. Use negative value to disable the health checking mechanism.
+         * @return {@code this}.
+         */
+        public RoundRobinLoadBalancerFactory.Builder<ResolvedAddress, C> healthCheckFailedConnectionsThreshold(
+                int threshold) {
+            this.healthCheckFailedConnectionsThreshold = threshold;
+            return this;
+        }
+
         /**
          * Builds the {@link RoundRobinLoadBalancerFactory} configured by this builder.
          *
          * @return a new instance of {@link RoundRobinLoadBalancerFactory} with settings from this builder.
          */
         public RoundRobinLoadBalancerFactory<ResolvedAddress, C> build() {
-            return new RoundRobinLoadBalancerFactory<>(eagerConnectionShutdown);
+            if (this.healthCheckFailedConnectionsThreshold < 0) {
+                return new RoundRobinLoadBalancerFactory<>(eagerConnectionShutdown, null);
+            }
+            assert !healthCheckInterval.isNegative();
+
+            if (this.backgroundExecutor == null && this.healthCheckInterval.isZero()
+                    && this.healthCheckFailedConnectionsThreshold == 0) {
+                return new RoundRobinLoadBalancerFactory<>(eagerConnectionShutdown,
+                        SharedHealthCheckConfig.getInstance());
```

Comment:
I would say this `if` and `SharedHealthCheckConfig` are unnecessary. Just check the executor and always allocate a new `HealthCheckConfig`.

Reply:
Sure.